R/methods-epi_df.R

```diff
@@ -80,13 +80,39 @@ summary.epi_df = function(object, ...) {
   } 
   
   cn <- names(res)
-  nr <- vctrs::vec_size(x) 
-  not_epi_df <- (!("time_value" %in% cn) || !("geo_value" %in% cn) || vctrs::vec_size(res) > nr || any(i > nr))
   
-  if (not_epi_df) return(tibble::as_tibble(res))
+  # Duplicate columns, Abort
+  dup_col_names = cn[duplicated(cn)]
+  if (length(dup_col_names) != 0) {
+    Abort(paste0("Column name(s) ", 
+                 paste(unique(dup_col_names), 
+                       collapse = ", "), " must not be duplicated."))
+  } 
+  
+  # If not an epi_df (missing time_value or geo_value), return a tibble
+  not_epi_df <- (!("time_value" %in% cn) || !("geo_value" %in% cn))
   
-  # Use reclass as safeguard (in case class & metadata are dropped)
+  if (not_epi_df) {
+    attributes(res)$metadata <- NULL 
+    return(tibble::as_tibble(res))
+  }
+  
+  # Use reclass as safeguard (in case class &/or metadata are dropped)
   reclass(res, attr(x, "metadata"))
+  
+  # Amend additional metadata if some other_keys cols are dropped
+  
+  # Extract the other_keys colnames in the original epi_df that do not occur in the subset
+  only_in_orig = setdiff(unlist(attr(x, "metadata")$other_keys), 
+                         cn[!(cn %in% c("geo_value", "time_value"))]) 
+  
+  if (length(only_in_orig) != 0) {
+    # remove all other_keys elements in the original epi_df that do not occur in subset
+    attr(res, "metadata")$other_keys = 
+      attr(res, "metadata")$other_keys[! attr(res, "metadata")$other_keys %in% only_in_orig]
+  }
```

Comment:
I think (since we're not messaging about this inside the `if` statement) this section might be able to be written shorter as
```
attr(res, "metadata")$other_keys <- intersect(attr(res, "metadata")$other_keys, cn)
```
and I believe this preserves the original order of the keys that stick around, but `?intersect` doesn't seem to guarantee this.  So if we want to guarantee this behavior, maybe
```
attr(res, "metadata")$other_keys <- attr(res, "metadata")$other_keys[attr(res, "metadata")$other_keys %in% cn]
```
plus a comment about why we're not using `intersect`?

Reply:
After looking into `intersect`, it looks like the order of the resulting elements is the same as the order of the elements in the 1st argument. So, we can use that. Also, `cn = names(res)`... so I think what you meant for the 1st argument in `intersect` is is `attr(x, "metadata")$other_keys` (b/c one of the args should be that)?

Reply:
Right, `intersect` in my version of R also appears to keep the first arg's order; however, I can't find anywhere in the docs that guarantees that this will be the case for any version of R.  That's why I was still looking at that `%in%` solution; in case for some reason they decide to change `intersect`'s implementation within the bounds of the docs.

I think I wrote it assuming that `attr(res, "metadata")` was set to `attr(x, "metadata")`; if that's not the case, then yes, sounds like something should be changed to `x` in there.

Reply:
Pushing an update to address this wild fear + to simplify a little bit (we don't even need the if statement!)